src/src/com/microsoft/aad/adal/HttpWebRequest.java

```diff
@@ -107,107 +89,103 @@ private void setupConnection() {
-                // this work for different SDKs
-                getStatusCode(response);
-
-                if (responseStream != null) {
-
-                    ByteArrayOutputStream byteStream = new ByteArrayOutputStream();
-                    byte[] buffer = new byte[4096];
-                    int bytesRead = -1;
-
-                    // Continue to read from stream if not cancelled and not EOF
-                    while ((bytesRead = responseStream.read(buffer)) > 0) {
-                        byteStream.write(buffer, 0, bytesRead);
-                    }
-
-                    responseBody = byteStream.toByteArray();
-                }
+                responseBody = byteStream.toByteArray();
+            }
 
-                // It will only run in debugger and set from outside for testing
-                if (android.os.Debug.isDebuggerConnected() && sDebugSimulateDelay > 0) {
-                    // sleep background thread in debugging mode
-                    Logger.v(TAG, "Sleeping to simulate slow network response");
+            // It will only run in debugger and set from outside for testing
+            if (Debug.isDebuggerConnected() && sDebugSimulateDelay > 0) {
+                // sleep background thread in debugging mode
+                Logger.v(TAG, "Sleeping to simulate slow network response");
+                try {
                     Thread.sleep(sDebugSimulateDelay);
+                } catch (InterruptedException e) {
+                    Logger.v(TAG, "Thread.sleep got interrupted exception " + e);
                 }
-
-                Logger.v(TAG, "Response is received");
-                response.setBody(responseBody);
-                response.setResponseHeaders(mConnection.getHeaderFields());
-            } catch (InterruptedException e) {
-                Logger.v(TAG, "Thread.sleep got interrupted exception " + e);
-            } catch (IOException e ) {
-                Logger.e(TAG, "IOException:" + e.getMessage(), " Method:" + mRequestMethod,
-                        ADALError.SERVER_ERROR, e);
-                mException = e;
-            } finally {
-                mConnection.disconnect();
-                mConnection = null;
             }
+
+            Logger.v(TAG, "Response is received");
+            response = new HttpWebResponse(statusCode, responseBody, connection.getHeaderFields());
+        } finally {
+            connection.disconnect();
         }
 
-        response.setResponseException(mException);
         return response;
     }
 
-    private void getStatusCode(HttpWebResponse response) throws IOException {
+    private int getStatusCode(HttpURLConnection connection) throws IOException {
         int statusCode = HttpURLConnection.HTTP_BAD_REQUEST;
 
         try {
-            statusCode = mConnection.getResponseCode();
+            statusCode = connection.getResponseCode();
         } catch (IOException ex) {
 
             if (Build.VERSION.SDK_INT < 16) {
```

Comment:
Nope, right after...but parsing what in input streams comes after status code check